_scala3-reference/language-versions/binary-compatibility.md

```diff
@@ -0,0 +1,38 @@
+---
+title: Binary Compatibility
+type: section
+num: 102
+previous-page: /scala3/reference/language-versions/source-compatibility
+---
+
+In Scala 2 different minor versions of the compiler were free to change the way how they encode different language features in JVM bytecode so each bump of the compiler's minor version resulted in breaking binary compatibility and if a project had any Scala dependencies they all needed to be (cross-)compiled to the same minor Scala version that was used in that project itself. On the contrary, Scala 3 has a stable encoding into JVM bytecode.
```

Comment:
Minor suggestions:

```suggestion
In Scala 2, different minor versions of the compiler were free to change the way they encoded different language features in JVM bytecode. So, each bump of the compiler's minor version resulted in breaking the binary compatibility, and if a project had any Scala dependencies they all needed to be (cross-)compiled to the same minor Scala version that was used in that project itself. On the contrary, Scala 3 has a stable encoding into JVM bytecode.
```

However, I am not sure we want to guarantee that “Scala 3 has a stable encoding into JVM bytecode”? Has this been discussed previously? I thought that, at some point we would like to leverage new features of the JVM, and we could recompile TASTy artifacts to still be able to use Scala libraries compiled before those new features of the JVM were used by the compiler, without having to recompile everything. But, I am not sure what are the plans now.

Reply:
https://github.com/lampepfl/dotty/pull/14156#discussion_r786765757